### ISSUE TITLE
Duplicity hangs trying to upload files to remote server

### ISSUE THREAD

**Issue by masterito** [paramiko/paramiko#331]
[Maintainer note: this thread kinda got hijacked by the maybe-related, maybe-not-related Duplicity problem mentioned in comments below. This OP is left as an archeological site. Sorry OP.]

I can connect to the host without issues but command execution doesn't work as expected:

In [13]: ssh.get_transport().is_alive()
Out[13]: True

In [14]: stdin, stdout, stderr = ssh.exec_command('ls')
DEBUG:paramiko.transport:[chan 2] Max packet in: 34816 bytes
DEBUG:paramiko.transport:[chan 2] Max packet out: 32768 bytes
INFO:paramiko.transport:Secsh channel 2 opened.
DEBUG:paramiko.transport:[chan 2] Sesch channel 2 request ok

In [15]: stdout.channel.recv_ready()
Out[15]: False

In [16]: stdout.channel.exit_status_ready()
Out[16]: False

On the other hand:
In [5]: chan = ssh.invoke_shell()

In [6]: chan.send('uname')
Out[6]: 5

In [7]: chan.recv_ready()
Out[7]: True

In [8]: chan.exit_status_ready()
Out[8]: False

In [9]: chan.send('\n')
Out[9]: 1

In [10]: chan.exit_status_ready()
Out[10]: False

In [11]:

In [11]: chan.recv_ready()
Out[11]: True

In [13]: chan.recv(1024)
Out[13]: 'Last login: Sat Apr 26 04:58:23 2014 from XXX\r\r\n[USER@XXX ~]# uname\r\nLinux\r\n[USER@XXX ~]# '

So command was executed successfully but paramiko doesn't know about it.

**Comment by phaidros7** [comment]
I can confirm the bug and the solution/hack from @nutztherookie to solve it on debian jessie with:
ii  duplicity                      0.6.24-1                  amd64        encrypted bandwidth-efficient backup
ii  python-paramiko                1.15.1-1                  all          Make ssh v2 connections with Python (Python 2)

**Comment by sebastianw** [comment]
I have the same issue with paramiko. One-Liner from @nutztherookie works for me, too.

**Comment by mika** [comment]
The workaround mentioned by @nutztherookie (thanks for that!) fixes the problem also for me. Since this problem is visible only with an increased verbosity level at all and even aborting the duplicity job via ctrl-c doesn't really work it would be great to get a fix for paramiko into Debian/jessie (with my Debian developer hat on). Is there any expected side-effect by just removing that line overall?

**Comment by bitprophet** [comment]
I do worry that simply removing that line will cause side effects of some kind - presumably it was put in place for a reason (though it far predates my involvement).

From a quick glance (and again, I am NOT previously familiar with this chunk of the code) it feels like the issue is both @nutztherookie's observation re: window size, combined with https://github.com/paramiko/paramiko/blob/7a6e89bcbd75c495205ddf7520c044000c2e6a65/paramiko/channel.py#L1171 never being happy about the state of `self.closed` or `self.eof_sent` - possibly something else in the state machine involved is never tickling those, or is never doing so prior to the timeout.

What I'd like myself or someone to do before "just merging" the suggested fix is:

* Check up on the git history and/or protocol spec re: why we do this window-wait at all (given that removing it doesn't seem to cause immediate explosions for people) - maybe other changes (such as the window size one) have obviated the need for it?
* Figure out why the closed/EOF-sent flags are not getting set as the wait method expects. Is it a too-aggressive timeout or something else?

**Comment by tholu** [comment]
I have the same problem using duply with duplicity on a recently updated Debian jessie.

**Comment by bitprophet** [comment]
Trying to get a repro case for this now, I don't think anybody has actually stated any details about their local use case besides "was using duplicity" :( (besides the OP, who similarly didn't give many details besides "was running a command").

Tried grabbing duplicity as a contributor/developer, but oof, it's not your average Python app (and 0.7 changed a lot, and 0.6 can't see my librsync). Nor can I install it successfully via homebrew.

Installed it on a Debian 8 VM, got Duplicity 0.6.24 and Debian's packaged Paramiko (1.15.1), same as an earlier commenter.

As I was worried would happen: can't reproduce trivially. E.g. with SFTP enabled on my local sshd, I can run `duplicity <random folder of stuff> scp://localhost//tmp/` and it works fine, no hang, no error, `/tmp/` gets populated with happy little GPG files.

Happy to try poking this as I noted last time around, but I need a reproduce case first :( cc: everyone.

**Comment by alexandrnikitin** [comment]
Not sure how it can help but I encountered the same issue and seems "solved" it http://stackoverflow.com/a/38789910/974487

**Comment by loswochos** [comment]
Took python-paramiko from the unstable tree on a debian jessie , works for me (https://packages.debian.org/de/sid/python-paramiko).

**Comment by prehensilecode** [comment]
Can confirm that @nutztherookie 's fix works for paramiko-2.0.2

**Comment by jhuebner79** [comment]
Had to hack around it like described by @nutztherookie for current Debian stable. No problems using Gentoo or Ubuntu.

**Comment by bitprophet** [comment]
Took another gander at this code-wise since I doubt I'd be able to reproduce if I couldn't before. As of the 1.17 branch:

- `Channel.out_window_size` starts initialized to zero
- It's likely most often first set to nonzero via `Channel._set_remote_channel` which is executed whenever a channel open success/confirmation message is received (e.g. https://tools.ietf.org/html/rfc4254#section-5.1 - this is setting the channel's outbound flow control window)
- It can be bumped up by receiving a window-adjust message inside `Transport`, which calls `Channel._window_adjust`. The RFC doesn't specify when these must/may/would normally be sent, so it's presumably implementation-specific.
- It can be shrunk within `Channel._wait_for_send_window` itself, i.e. "we're sending N bytes, so the window open to us is now N bytes smaller until we get an adjustment message."

I see no other adjustments to this value.

----

Revisiting @nutztherookie's assertions from years ago (sob):

> Apparently in version <= 09e9d48 the "self.out_window_size == 0" is NEVER True (the size stays somewhere around 4M, I think)
> In newer versions self.out_window_size becomes 0 eventually and the method times out in that if-block.

The latter squares with the above analysis (presumably in older code the window size was not correctly managed, aka, a bug?) re: window size shrinking to 0. Then, because this is how windows work, the code is supposed to wait around, hoping that something else out-of-band raises the window size, and if nothing does, we must timeout.

Going by my above analysis, the only way that this would ever _not_ time out is if the remote end sends us a window-adjust message ([SSH_MSG_CHANNEL_WINDOW_ADJUST](https://tools.ietf.org/html/rfc4254#section-5.2)). Notably, the `self.out_buffer_cv.wait(timeout)` call within `_wait_for_window_size`'s loop is a `threading.Condition` (on `self.lock`) that allows this loop to release `self.lock` until it's notified, which occurs on either window adjustment or channel closure.

----

So - at that level, this is all Working As Intended, meaning one possible cause is a misbehaving server not sending window adjustments when we'd really rather it did; or some problem preventing us from processing those messages correctly (seems less likely given how simple that part of the networking is?)

Double checking other angles on this chunk of the code, what else could be breaking, such that not calling `_wait_for_window_size` is somehow a "fix" (with...no clearly defined symptoms yet)?

- TK

----

Unfortunately, nobody who's confirmed that "the fix works" has **told me how to reliably reproduce the problem** and now that I understand this better, **I am definitely not happy just turning off window management entirely**. It's part of the protocol!

**Comment by ploxiln** [comment]
tl;dr if anyone experiences this problem and removing that one line works, please report:

  * python version
  * paramiko version
  * client system OS version (e.g. "Ubuntu 16.04.1 kernel 4.4.0-62-generic")
  * client system hardware type (virtual machine? laptop model/year? server cpu/mem?)
  * specific server system OS version
  * server system hardware type
  * ssh server version (e.g. "OpenSSH_7.2p2 Ubuntu-4ubuntu2.1, OpenSSL 1.0.2g")
  * ping time (latency) between client and server, throughput (rate) when using plain ssh
  * the program you were using (duplicity?) and where you got it (standard repo?)
  * how regularly the problem happens (every time?)

**Comment by bitprophet** [comment]
I'm seeing Duplicity calling `settimeout` in its Paramiko backend, whose value appears to default to `30` - so that's likely ruled out.

Only leaves me with, as usual, questions about the detailed situations where people are getting this. Local platform & Python version, remote server & version, etc. Clearly it's not affecting all Paramiko SFTP usage or I'd have heard much more about it or encountered it myself. So I'm still guessing there's some pattern linking everyone this impacts.

Other brainstorming:

- Window adjustment needs to hold the lock, so if there was some bug in Python's threading that prevented the threading condition from working the way it's documented - and thus window adjust method and our data send method were deadlocked - that could be one cause. Seems quite unlikely to me, for obvious reasons, but Python core bugs have been a thing before.
    - It looks like folks can manually adjust the constant `self.ultra_debug = False` in `Channel` to be `True` if they want to see debug log messages when the window size changes...FYI.
- If the window adjust message contained an int of `0`, that would also be kind of problematic as we'd not be deadlocked but we'd also never grow our window size and thus never escape the loop. Also seems rather unlikely.
- If `Channel._close_internal` is called but its opening `if not self.active or self.closed` clause is tripped already, it would `return` and never get to `self._set_closed()`, which is the only other way to trigger the condition our window waiter is waiting on. There have been numerous other tickets about problems surrounding session teardown, so this doesn't seem out of the realm of possibility.

Yet again, all of this requires either:

- someone to help me reproduce this reliably on my end (ideally _without_ Duplicity directly involved, but since I can apparently install it OK on Linux, if that's what it takes, that's fine)
- someone to dig into any of the above ideas on their own

**Comment by firlevapz** [comment]
Hi, i can reproduce the problem with duplicity, here are the details of my system and I'm happy to provide further information, if it helps, but I don't really understand the root cause of it. The fix of @nutztherookie to comment out https://github.com/paramiko/paramiko/blob/master/paramiko/channel.py#L1059 helped.

- python 2.7.9
- python-paramiko 1.15.1-1 from debian jessie packages
- duplicity 0.6.24-1 from debian jessie packages
- Linux srv-01 4.4.59-1-pve #1 SMP PVE 4.4.59-87 (Proxmox cluster) at a physical server machine
- ping ~ 0.3 ms to sftp server
- Openssh-Server: OpenSSH_6.7p1 Debian-5+deb8u3, SSH-2.0-mod_sftp/0.9.9
- happens every time

**Comment by stefangweichinger** [comment]
Same situation as @firlevapz (deb 8.8, packages from jessie), target system is a Hetzner Storage Box.

Line 1059 in /usr/lib/python2.7/dist-packages/paramiko/channel.py looks different here, pls advise on how to work around this issue. Am I looking at the wrong file?

**Comment by ploxiln** [comment]
@firlevapz linked to master, which continued to change. @stefangweichinger the line in question is https://github.com/paramiko/paramiko/blob/v1.15.1/paramiko/channel.py#L1059 (link is specifically do v1.15.1 tag instead of master)

The problem seems to be with the "window", a concept similar to the TCP "receive window". It is still a bit unclear exactly where the problem is (possibly with the remote server) - just removing this would presumably make paramiko work worse for other purposes.

**Comment by stefangweichinger** [comment]
thanks, now it works with duplicity.

btw: rolling back also made me hit this one again:
https://dailydebugtechlove.blogspot.cz/2017/02/how-to-fix-paramiko-issue-valueerror.html 
I know, other topic/bug, but anyway: I had to apply that fix as well.

**Comment by SuperTux88** [comment]
@stefangweichinger I had the same problem with jessie and hetzner storagebox, but I used scp://. After upgrading to stretch, I needed to switch to sftp:// because with scp:// it can't create directories with the stretch version of duplicity (so I never tested with jessie and sftp://, but I see it doesn't work for you). And with stretch and sftp:// everything works perfectly, without modifying anything at the code (also your second comment works with stretch). So whatever it was, it seems fixed in debian stretch 👍

**Comment by stefangweichinger** [comment]
@SuperTux88 thanks for the feedback. scp:// never worked for me in this case, even with jessie.
I had waited with upgrade to stretch as I have to upgrade 2 "identical" machines (test and prod), but after thinking about it once more I remembered being able to snapshot the test-machine before and really *test* things that way ;-) stretch upgrade runs right now. will report.

**Comment by okurz** [comment]
Changing the path to sftp on jessie also did not help for me to make use of duplicity. I went with changing the backend to pexpect as mentioned in https://github.com/paramiko/paramiko/issues/331#issuecomment-48108533 , i.e. `duplicity --ssh-backend=pexpect …` . Sorry I can't provide help to fix paramiko.

**Comment by func0der** [comment]
Workaround: Use sshfs, mount the storage and copy to the mount path via `file://`, if you do not want to work on the code.

**Comment by cruizer** [comment]
Probably not the same root cause, but the symptoms surely look very similar, so putting this here in case someone also stumbles on this GH issue:

In my case, when starting a backup from scratch in Duplicity using the Paramiko backend for SCP backup, Paramiko was throwing a timeout exception from here: https://github.com/paramiko/paramiko/blob/master/paramiko/channel.py#L701

Then I found this bug report https://bugs.launchpad.net/duplicity/+bug/1859304 and applying the patch shared in it has solved my issue.

- Duplicity version:  0.8.9
- Paramiko version: 2.7.1
- Python 3.6.8
- OS: CentOS 8.1.1911
